apiserver/facades/agent/uniter/register.go

```diff
@@ -93,12 +101,18 @@ func newUniterAPIWithServices(
 
 	accessUnitOrApplication := common.AuthAny(accessUnit, accessApplication)
 
+	modelInfo, err := modelInfoService.GetModelInfo(stdCtx)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	modelTag := names.NewModelTag(modelInfo.UUID.String())
+
 	cloudSpec := cloudspec.NewCloudSpecV2(resources,
 		cloudspec.MakeCloudSpecGetterForModel(st, cloudService, credentialService),
 		cloudspec.MakeCloudSpecWatcherForModel(st, cloudService),
 		cloudspec.MakeCloudSpecCredentialWatcherForModel(st),
 		cloudspec.MakeCloudSpecCredentialContentWatcherForModel(st, credentialService),
-		common.AuthFuncForTag(m.ModelTag()),
+		common.AuthFuncForTag(modelTag),
```

Comment:
This is bugging me a bit. Probably nothing we can do about it at the moment. But I don't like that this can't be late bound or the model info service passed to it. Going to circulate the problem.

Reply:
I think the `NewCloudSpecV2` (I despise that name) should take in real services rather than funcs to funcs with late binding. Just operate on concrete things. Most of the `Make` functions aren't making anything, they're just wrappers around calling out to other functions (`MakeCloudSpecWatcherForModel`, `MakeCloudSpecCredentialWatcherForModel`, and `MakeCloudSpecCredentialContentWatcherForModel`).

Reply:
As @manadart says, this is out of scope for this PR, but I've added it to the risk register and we can improve this in future work